File: lib/src/main/java/io/xj/lib/meme/ParseStrong.java
// Copyright (c) XJ Music Inc. (https://xjmusic.com) All Rights Reserved.

package io.xj.lib.meme;

import java.util.List;
import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Meme Matcher for Strong-Memes
 * <p>
 * Parse any meme to test if it's valid, and extract its features
 * <p>
 * Strong-meme like LEMONS! should always favor LEMONS https://www.pivotaltracker.com/story/show/180468772
 */
class ParseStrong {
  static final Pattern rgx = Pattern.compile("^(.+)!$");
  final String body;
  final boolean isValid;

  ParseStrong(String raw) {
    Matcher matcher = rgx.matcher(raw);

    if (!matcher.find()) {
      body = raw;
      isValid = false;
      return;
    }

    body = matcher.group(1);
    if (java.util.Objects.isNull(body) || body.length() == 0) {
      isValid = false;
      return;
    }

    isValid = true;
  }

  public static ParseStrong fromString(String raw) {
    return new ParseStrong(raw);
  }

  public boolean isAllowed(List<ParseStrong> memes) {
    if (!isValid) return true;
    for (var meme : memes)
      if (Objects.equals(body, meme.body)) return true;
    return false;
  }
}

File: engine/src/meme/ParseStrong.cpp
// Copyright (c) XJ Music Inc. (https://xjmusic.com) All Rights Reserved.

#include "xjmusic/meme/ParseStrong.h"

using namespace XJ;

ParseStrong::ParseStrong(const std::string &raw) {
  std::smatch matcher;
  valid = std::regex_search(raw, matcher, rgx);

  if (!valid) {
    body = raw;
    valid = false;
    return;
  }

  body = matcher[1].str();
  if (body.empty()) {
    valid = false;
    return;
  }

  valid = true;
}

ParseStrong ParseStrong::fromString(const std::string &raw) {
  return ParseStrong(raw);
}

bool ParseStrong::isAllowed(const std::vector<ParseStrong> &memes) const {
  if (!valid) return true;
  return !std::all_of(memes.begin(), memes.end(), [this](const ParseStrong &meme) {
    return body != meme.body;
  });
}

const std::regex ParseStrong::rgx("^(.+)!$");
